plan/point_select_plan.go

```diff
@@ -0,0 +1,435 @@
+func (fp *PointSelectPlan) ResolveIndices() {}
+
+func tryFastPlan(ctx sessionctx.Context, node ast.Node) Plan {
+	if PreparedPlanCacheEnabled {
+		// plan cache not supported yet.
+		return nil
+	}
+	switch x := node.(type) {
+	case *ast.SelectStmt:
+		fp := tryFastSelectPlan(ctx, x)
+		if fp != nil {
+			if checkFastPlanPrivilege(ctx, fp, mysql.SelectPriv) != nil {
+				return nil
+			}
+			return fp
+		}
+	case *ast.UpdateStmt:
+		return tryUpdatePointPlan(ctx, x)
+	case *ast.DeleteStmt:
+		return tryDeletePointPlan(ctx, x)
+	}
+	return nil
+}
+
+func tryFastSelectPlan(ctx sessionctx.Context, selStmt *ast.SelectStmt) *PointSelectPlan {
+	if selStmt.GroupBy != nil || selStmt.Having != nil || selStmt.OrderBy != nil || selStmt.Limit != nil ||
+		selStmt.LockTp != ast.SelectLockNone {
+		return nil
+	}
+	tblName := getSingleTableName(selStmt.From)
+	if tblName == nil {
+		return nil
+	}
+	tbl := tblName.TableInfo
+	if tbl == nil {
+		return nil
+	}
+	for _, col := range tbl.Columns {
+		// Do not handle generated columns.
+		if col.IsGenerated() {
+			return nil
+		}
+		// Only handle tables that all columns are public.
+		if col.State != model.StatePublic {
+			return nil
+		}
+	}
+	pairs := make([]nameValuePair, 0, 4)
+	pairs = getNameValuePairs(pairs, selStmt.Where)
+	if pairs == nil {
+		return nil
+	}
+	handleDatum := findPKHandle(tbl, pairs)
+	if handleDatum.Kind() == types.KindInt64 {
+		schema := buildSchemaFromFields(ctx, tbl, selStmt.Fields.Fields)
+		if schema == nil {
+			return nil
+		}
+		p := newPointSelectPlan(ctx, schema, tbl)
+		p.Handle = handleDatum.GetInt64()
+		return p
+	}
+	for _, idxInfo := range tbl.Indices {
+		if !idxInfo.Unique {
+			continue
+		}
+		if idxInfo.State != model.StatePublic {
+			return nil
+		}
+		idxValues := getIndexValues(idxInfo, pairs)
+		if idxValues == nil {
+			continue
+		}
+		schema := buildSchemaFromFields(ctx, tbl, selStmt.Fields.Fields)
+		if schema == nil {
+			return nil
+		}
+		p := newPointSelectPlan(ctx, schema, tbl)
+		p.IndexInfo = idxInfo
+		p.IndexValues = idxValues
+		return p
+	}
+	return nil
+}
+
+func newPointSelectPlan(ctx sessionctx.Context, schema *expression.Schema, tbl *model.TableInfo) *PointSelectPlan {
+	p := &PointSelectPlan{
+		basePlan: newBasePlan(ctx, "Point_Select"),
+		schema:   schema,
+		TblInfo:  tbl,
+	}
+	p.stats = &statsInfo{count: 1}
+	return p
+}
+
+func checkFastPlanPrivilege(ctx sessionctx.Context, fastPlan *PointSelectPlan, checkTypes ...mysql.PrivilegeType) error {
+	pm := privilege.GetPrivilegeManager(ctx)
+	if pm == nil {
+		return nil
+	}
+	dbName := ctx.GetSessionVars().CurrentDB
+	for _, checkType := range checkTypes {
+		if !pm.RequestVerification(dbName, fastPlan.TblInfo.Name.L, "", checkType) {
+			return errors.New("privilege check fail")
+		}
+	}
+	return nil
+}
+
+func buildSchemaFromFields(ctx sessionctx.Context, tbl *model.TableInfo, fields []*ast.SelectField) *expression.Schema {
+	dbName := model.NewCIStr(ctx.GetSessionVars().CurrentDB)
```

Comment:
Why current db?

Reply:
We should pass in the DB name.

Reply:
But it is not necessarily the current db, for example, `test.t`.